sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubProducerClient.cs

```diff
@@ -419,22 +445,40 @@ public virtual Task<string[]> GetPartitionIdsAsync(CancellationToken cancellatio
                 // its idle timeout period elapses.
 
                 Argument.AssertNotClosed(IsClosed, nameof(EventHubProducerClient));
-                activeProducer = PartitionProducers.GetOrAdd(options.PartitionId, id => Connection.CreateTransportProducer(id, RetryPolicy));
-            }
 
-            using DiagnosticScope scope = CreateDiagnosticScope();
+                bool isMessageSent = false;
+                int failedAttemptCount = 0;
 
-            events = events.ToList();
-            InstrumentMessages(events);
+                using DiagnosticScope scope = CreateDiagnosticScope();
+                events = events.ToList();
+                InstrumentMessages(events);
 
-            try
-            {
-                await activeProducer.SendAsync(events, options, cancellationToken).ConfigureAwait(false);
-            }
-            catch (Exception ex)
-            {
-                scope.Failed(ex);
-                throw;
+                while (!isMessageSent)
+                {
+                    var pooledProducer = PartitionProducerPool.GetPartitionProducer(options.PartitionId, Connection, RetryPolicy, options.RemoveAfterDuration);
```

Comment:
I'd prefer to see us consider doing this in the "get me the right producer" area and not fork the entire set of logic.    It may be worthy of extracting into a dedicated get-me-a-producer method.   I'm thinking something more like:

```csharp

// No need for the retry policy here; this is a very limited and special case.
// We should never hit this more than once, so give it a tiny bit of leeway, just in case.   
// I could also be talked into 5 here.

private const MaximumCreateProducerAttempts = 3;
private static readonly PartitionProducerLifespan = TimeSpan.FromMinutes(5);

private TransportProducer GetTransportProducer(string partitionId, CancellationToken cancellationToken)
{
    if (string.IsNullOrEmpty(partitionId)
    {
        return EventHubProducer;  
    }
      
    TransportProducer producer = null;
    int attempts = 0;

    while ((!cancellationToken.IsCancellationRequested) && (producer == null)
    {
            producer = PartitionProducerPool.GetPartitionProducer(partitionId, Connection, RetryPolicy, PartitionProducerLifespan).Item;

            // Only try again if the producer client and connection aren't closed.
            if ((producer.IsClosed) && (!IsClosed) && (!Connection.IsClosed))
            {
                    if (++attempts > MaximumCreateProducerAttempts)
                    {
                          throw new EventHubsException(EventHub,  EventHubsException.FailureReason.ClientClosed);
                    }

                    producer = null;
            }
    }

    // If we get here, we haven't thrown, so return the sucker.
    return producer;
}
```

Reply:
note:  From earlier in the conversation, I see that I was the genius that was suggesting we consider using the retry policy.   I've gone and said some very unkind things to myself while looking in the mirror as recompense.   Apologies for the churn.

Reply:
Thinking about this more, this isn't likely to catch the full race condition;  we'd need similar logic around the send operation itself.   Something like:

```c#

try
{
    var attempts = 0;

    while (++attempts <= MaximumCreateProducerAttempts)
    {
        try
        {
            await activeProducer.SendAsync(events, options, cancellationToken).ConfigureAwait(false);
        }
        catch (EventHubsException ex) when (eventHubException.Reason == EventHubsException.FailureReason.ClientClosed)
        {
              if ((activeProducer.IsClosed) && (!IsClosed) && (!Connection.IsClosed))
              {
                  activeProducer = GetTransportProducer(partitionId, cancellationToken);
              }
              else
              {
                  throw;
              }
        }
    }
}
catch (Exception ex)
{
    scope.Failed(ex);
    throw;
}
```

It's a bit of duplication.   I'm on the fence if we should not bother with doing it in the potential `GetTransportProducer` method.   Definitely open to some thoughts here.

Reply:
> Apologies for the churn.

It was looking like a good idea to me, sorry for that.

Reply:
> Definitely open to some thoughts here.

I should come back to you soon on this.

Reply:
Not at all;  Past Jesse owes us both an apology for half-baked thoughts and bad advice.   I appreciate you being flexible.   Its frustrating when someone flip-flops on you after you've done a bunch of work, and I regret doing that to you.

Reply:
hrmm.... we probably could compact that a bit and remove some of the duplication with a small helper.   Something like:

```csharp
try
{
    var attempts = 0;

    while (++attempts <= MaximumCreateProducerAttempts)
    {
        try
        {
            await activeProducer.SendAsync(events, options, cancellationToken).ConfigureAwait(false);
        }
        catch (EventHubsException ex) when 
            ((eventHubException.Reason == EventHubsException.FailureReason.ClientClosed) && (ShouldRecreateProducer(activeProducer));
        {
              activeProducer = GetTransportProducer(partitionId, cancellationToken);
        }
    }
}
catch (Exception ex)
{
    scope.Failed(ex);
    throw;
}
```
```csharp
private bool ShouldRecreateProducer(TransportProducer producer) => ((producer.IsClosed) && (!IsClosed) && (!Connection.IsClosed));
```

Confession:  I'm really not in love with the name `ShouldRecreateProducer`, but I'm struggling to come up with something better.   Thoughts are definitely appreciated.

Reply:
> Definitely open to some thoughts here.

I wonder how would look to you something like this.

Split the call to ```PartitionProducerPool.GetPartitionProducer``` in two calls:
- ```PartitionProducerPool.GetPooledProducer```: It would set up the tracking of the ```PoolItem``` in the pool
-  ```PooledProducer.GetTransportProducer```: It would retrieve/create the ```TransportProducer``` from a ```PoolItem```

```csharp
await using (var pooledProducer = PartitionProducerPool.GetPooledProducer(options.PartitionId, options.RemoveAfterDuration))
{
    TransportProducer activeProducer = pooledProducer.GetTransportProducer(Connection, RetryPolicy);

    using DiagnosticScope scope = CreateDiagnosticScope();

    events = events.ToList();
    InstrumentMessages(events);

    try
    {
        await activeProducer.SendAsync(events, options, cancellationToken).ConfigureAwait(false);
    }
    catch (Exception ex)
    {
        scope.Failed(ex);
        throw;
    }
}

```

I removed the retry logic to make it more readable.

Reply:
> Its frustrating when someone flip-flops on you after you've done a bunch of work, and I regret doing that to you.

No worries, it was great that you realized of how to improve it.

Reply:
Fair point, I hadn't thought about the disposal.   I think we can merge the two approaches by breaking things out a bit to reduce duplication and complexity.   This may be half-baked, so please consider it a conversation-starter:

```csharp
public virtual async Task SendAsync(EventDataBatch eventBatch, 
                                    CancellationToken cancellationToken = default)
{
    if (string.IsNullOrEmpty(eventBatch.PartitionId)
    {
       return PerformSend(EventHubProducer, eventBatch, cancellationToken);
    }

    var attempts = 0;

    while (++attempts <= MaximumCreateProducerAttempts)
    {
        var pooledProducer = GetTransportProducer(partitionId, cancellationToken);
        await using var producerAwaitable = activeProducer.ConfigureAwait(false);

        try
        {
            await PerformSend(pooledProducer.GetTransportProducer(Connection, RetryPolicy), eventBatch, cancellationToken).ConfigureAwait(false);
        }
        catch (EventHubsException ex) when 
            ((eventHubException.Reason == EventHubsException.FailureReason.ClientClosed) && (ShouldRecreateProducer(activeProducer));
        {
              // Do nothing; this is a retry case.
        }    
    }
}
```
```csharp
private async Task PerformSend(TransportProducer producer, 
                               EventDataBatch eventBatch, 
                               CancellationToken cancellationToken = default)
{
    Argument.AssertNotClosed(IsClosed, nameof(EventHubProducerClient));

    using DiagnosticScope scope = CreateDiagnosticScope();

    try
    {
        await producer.SendAsync(events, options, cancellationToken).ConfigureAwait(false);
    }
    catch (Exception ex)
    {
        scope.Failed(ex);
        throw;
    }
}

private async Task PerformSend(TransportProducer producer, 
                               IEnumerable<EventData> events,
                               SendEventOptions options,
                               CancellationToken cancellationToken = default)
{
    Argument.AssertNotClosed(IsClosed, nameof(EventHubProducerClient));

    using DiagnosticScope scope = CreateDiagnosticScope();

    events = events.ToList();
    InstrumentMessages(events);

    try
    {
        await producer.SendAsync(events, options, cancellationToken).ConfigureAwait(false);
    }
    catch (Exception ex)
    {
        scope.Failed(ex);
        throw;
    }
}
```
```csharp
private bool ShouldRecreateProducer(TransportProducer producer) => ((producer.IsClosed) && (!IsClosed) && (!Connection.IsClosed));
```

I'm only showing the public "send the batch" method, but I think the structure would be the same for both.    What I like about this is that it keeps the branched code very compact and separates the send operation from managing the producer.     Thoughts?

//cc: @kinelski -  _(could use your thoughts here as well(_